### DISCUSSION TITLE
How might we abstract S3?

### DISCUSSION THREAD

**Discussion by ceilfors** [nearst/laconia#933] [Ideas]
S3 is an essential part of a serverless architecture in AWS. We write and read a lot from S3. Laconia encourages technologists to not be dependent on the cloud in their core application. At the same time, I'm also reluctant to introduce an unnecessary abstraction. Is there a middle ground that covers 80% of user's use cases? How will the API look like?

**Comment by ljcoomber** [comment]
I have similar thoughts to #111 - essential avoiding premature abstraction in a framework for things that  are likely to require specific implementation detail.

That said, one use case I've seen recently would be triggering a Lambda on file change event from S3

**Comment by ceilfors** [comment]
Lambda trigger -- Laconia actually already supports this. See: https://laconiajs.io/docs/guides/adapting-events. It retrieves object automatically for you on Lambda trigger.

On premature abstraction, I can say that on most of my cases, my operations on S3 object operation. I'd like to think that we can abstract object operation properly i.e. not prematurely?

This is one example abstraction from [a tutorial from claudiajs](https://claudiajs.com/tutorials/designing-testable-lambdas.html):

```js
const fileSystem = require('./s3-file-system')

inputStream = fileSystem.readAsStream(event.filePath, event.location);

fileSystem.writeFromStream(
      outputFilePath, 
      event.location, 
      convertedStream, 
      callback
    ); 
```

Is a file system a good abstraction for us?

**Comment by hugosenari** [comment]
> Laconia encourages technologists to not be dependent on the cloud in their core application.

For today reality, S3 API become a 'de facto standard', almost every cloud provider [has a S3 compatible API](https://kopia.discourse.group/t/list-of-s3-compatible-cloud-storage-providers-that-work-with-kopia/1270/13).
Off course, this won't answer the question of 'how', but at least would give some ground to do it.
